scipy/signal/_upfirdn_apply.pyx

```diff
@@ -65,6 +65,56 @@ def _output_len(Py_ssize_t len_h,
     return need
 
 
+@cython.cdivision(True)  # faster modulo
+@cython.boundscheck(False)  # designed to stay within bounds
+@cython.wraparound(False)  # we don't use negative indexing
+def _rat(x, tol=None):
```

Comment:
You'll need to limit the number of iterations in the loop.  Calling `_rat(10000000001/10000000000.0)` never returns.  While its true that arrays of that size would be enormous, its still not good to have this possibility here. 

What constitutes a reasonable filter size? Are you okay with `_rat(10001/10000.0) = array([10001, 10000])`?  I think we may need to enforce some kind of limit more directly.